File: lib/chat_models/preloaded_action/preloaded_action.dart
import 'package:acs_plugin/chat_models/push_notification_chat_message_received_event/push_notification_chat_message_received_event.dart';
import 'package:acs_plugin/chat_models/utilities/map_helpers.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

part 'preloaded_action.freezed.dart';
part 'preloaded_action.g.dart';

/// Dart equivalent of the Swift enum PreloadedActionType
enum PreloadedActionType {
  @JsonValue('chatNotification')
  chatNotification,
}

@freezed
class PreloadedAction with _$PreloadedAction {
  const factory PreloadedAction({
    required PreloadedActionType type,
    @JsonKey(
        name: 'chatPushNotificationReceivedEvent', readValue: readValueObject)
    PushNotificationChatMessageReceivedEvent? chatPushNotificationReceivedEvent,
  }) = _PreloadedAction;

  factory PreloadedAction.fromJson(Map<String, dynamic> json) =>
      _$PreloadedActionFromJson(json);
}

File: ios/Classes/PreloadedAction.swift
//
//  PreloadedAction.swift
//  Pods
//
//  Created by Yriy Malyts on 22.05.2025.
//

import AzureCommunicationChat

enum PreloadedActionType: String {
    case chatNotification
}

struct PreloadedAction {
    let type: PreloadedActionType
    let chatPushNotificationReceivedEvent: PushNotificationChatMessageReceivedEvent?

    func toJson() -> [String: Any] {
        var json: [String: Any] = [
            "type": type.rawValue
        ]
        
        if let chatPushNotificationReceivedEvent = chatPushNotificationReceivedEvent {
            json["chatPushNotificationReceivedEvent"] = chatPushNotificationReceivedEvent.toJson()
        }
        
        return json
    }
}
